src/lib.rs

```diff
@@ -1278,203 +1290,360 @@ impl Format {
+        match next.as_ref() {
+            Next::Empty => Self::accept(),
+            Next::Union(next1, next2) => {
+                let tree1 = Self::add_next(module, next1.clone());
+                let tree2 = Self::add_next(module, next2.clone());
+                tree1.union(tree2)
+            }
+            Next::Cat(f, next) => Self::add(module, f, next.clone()),
+            Next::Tuple(fields, next) => Self::add_tuple(module, fields, next.clone()),
+            Next::Record(fields, next) => Self::add_record(module, fields, next.clone()),
+            Next::Repeat(a, next0) => {
+                let tree = Self::add_next(module, next0.clone());
+                tree.union(Self::add(module, a, next))
             }
-            Format::Fail => Ok(()),
-            Format::EndOfInput => self.accept(index),
-            Format::Align(_) => {
-                self.accept(index) // FIXME
+            Next::RepeatCount(n, a, next0) => Self::add_repeat_count(module, *n, a, next0.clone()),
+            Next::Slice(n, inside, next0) => {
+                Self::add_slice(module, *n, inside.clone(), next0.clone())
             }
-            Format::Byte(bs) => {
-                let mut bs = *bs;
-                let mut new_branches = Vec::new();
-                for (bs0, nexts) in self.branches.iter_mut() {
-                    let common = bs0.intersection(&bs);
-                    if !common.is_empty() {
-                        let orig = bs0.difference(&bs);
-                        if !orig.is_empty() {
-                            new_branches.push((orig, nexts.clone()));
-                        }
-                        *bs0 = common;
-                        nexts.add(index, next.clone())?;
-                        bs = bs.difference(bs0);
-                    }
-                }
-                if !bs.is_empty() {
-                    let mut nexts = Nexts::new();
-                    nexts.add(index, next.clone())?;
-                    self.branches.push((bs, nexts));
-                }
-                self.branches.append(&mut new_branches);
-                Ok(())
+            Next::Peek(next1, next2) => {
+                let tree1 = Self::add_next(module, next1.clone());
+                let tree2 = Self::add_next(module, next2.clone());
+                tree1.peek(tree2)
+            }
+            Next::PeekNot(next1, next2) => {
+                let tree1 = Self::add_next(module, next1.clone());
+                let tree2 = Self::add_next(module, next2.clone());
+                tree1.peek_not(tree2)
+            }
+        }
+    }
+
+    pub fn add(module: &'a FormatModule, f: &'a Format, next: Rc<Next<'a>>) -> MatchTreeStep<'a> {
+        match f {
+            Format::ItemVar(level, _args) => Self::add(module, module.get_format(*level), next),
+            Format::Fail => Self::reject(),
+            Format::EndOfInput => Self::accept(),
+            Format::Align(_) => {
+                Self::accept() // FIXME
```

Comment:
How difficult would it be to implement Align? Is it possible under the current model but just tedious, or is it actually not possible as designed?

Reply:
Align could be implemented by treating it as equivalent to some optional number of arbitrary bytes, but there's little value in doing that right now as we only use it in one place in the deflate format and it's not something likely to be switched upon.

It's similar to the slightly more likely case of the upper bound for RepeatCount, which is also ignored. For example if you have RepeatCount(expr, f) where the bounds for expr are min..max then currently we treat that as being min repetitions of f, but if max > min then we just stop at that point instead of exploring every possible value that max could take, as this isn't necessary for any format. Align(n) could be implemented as `RepeatCount(expr, byte)` where expr has bounds from 0..n-1 depending on the current offset.

Anyway I think we can ignore Align for now!